lib/src/palette_picker.dart

```diff
@@ -0,0 +1,184 @@
+    this.rightPosition = 1.0,
+    this.topPosition = 0.0,
+    this.bottomPosition = 1.0,
+    Key? key,
+  }) : super(key: key);
+
+  final Offset position;
+  final ValueChanged<Offset> onChanged;
+  final double leftPosition;
+  final double rightPosition;
+  final List<Color> leftRightColors;
+  final double topPosition;
+  final double bottomPosition;
+  final List<Color> topBottomColors;
+
+  @override
+  _PalettePickerState createState() => _PalettePickerState();
+}
+
+class _PalettePickerState extends State<PalettePicker> {
+  final GlobalKey paletteKey = GlobalKey();
+
+  Offset get position => super.widget.position;
+  double get leftPosition => super.widget.leftPosition;
+  double get rightPosition => super.widget.rightPosition;
+  double get topPosition => super.widget.topPosition;
+  double get bottomPosition => super.widget.bottomPosition;
+
+  /// Position(min, max) > Ratio(0, 1)
+  Offset positionToRatio() {
+    final double ratioX = leftPosition < rightPosition
+        ? positionToRatio2(position.dx, leftPosition, rightPosition)
+        : 1.0 - positionToRatio2(position.dx, rightPosition, leftPosition);
+
+    final double ratioY = topPosition < bottomPosition
+        ? positionToRatio2(position.dy, topPosition, bottomPosition)
+        : 1.0 - positionToRatio2(position.dy, bottomPosition, topPosition);
+
+    return Offset(ratioX, ratioY);
+  }
+
+  double positionToRatio2(
+      double postiton, double minPostition, double maxPostition) {
+    if (postiton < minPostition) return 0.0;
+    if (postiton > maxPostition) return 1.0;
+    return (postiton - minPostition) / (maxPostition - minPostition);
+  }
+
+  /// Ratio(0, 1) > Position(min, max)
+  void ratioToPosition(Offset ratio) {
+    final RenderBox? renderBox =
+        paletteKey.currentContext?.findRenderObject() as RenderBox?;
+    final Offset startposition =
+        renderBox?.localToGlobal(Offset.zero) ?? Offset.zero;
+    final Size size = renderBox?.size ?? Size.zero;
+    final Offset updateOffset = ratio - startposition;
+
+    final double ratioX = updateOffset.dx / size.width;
+    final double ratioY = updateOffset.dy / size.height;
+
+    final double positionX = leftPosition < rightPosition
+        ? ratioToPosition2(ratioX, leftPosition, rightPosition)
+        : ratioToPosition2(1.0 - ratioX, rightPosition, leftPosition);
+
+    final double positionY = topPosition < bottomPosition
+        ? ratioToPosition2(ratioY, topPosition, bottomPosition)
+        : ratioToPosition2(1.0 - ratioY, bottomPosition, topPosition);
+
+    final Offset position = Offset(positionX, positionY);
+    super.widget.onChanged(position);
+  }
+
+  double ratioToPosition2(
+    double ratio,
+    double minposition,
+    double maxposition,
+  ) {
+    if (ratio < 0.0) return minposition;
+    if (ratio > 1.0) return maxposition;
+    return ratio * maxposition + (1.0 - ratio) * minposition;
+  }
+
+  Widget buildLeftRightColors() {
+    return Container(
```

Comment:
If you got more time, replace `Container`s like this to a more precise widget, `DecoratedBox` here.